File: java/src/main/java/com/databricks/zerobus/ZerobusJsonStream.java
package com.databricks.zerobus;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

/**
 * Stream for ingesting JSON records into a table.
 *
 * <p>This class provides a clean API for JSON ingestion that doesn't require Protocol Buffer types.
 *
 * <p>Create instances using {@link ZerobusSdk#createJsonStream}:
 *
 * <pre>{@code
 * ZerobusJsonStream stream = sdk.createJsonStream(
 *     "catalog.schema.table",
 *     clientId,
 *     clientSecret
 * ).join();
 *
 * // Main: Ingest objects with a serializer
 * Gson gson = new Gson();
 * long offset = stream.ingestRecordOffset(myObject, gson::toJson);
 * stream.waitForOffset(offset);
 *
 * // Alt: Ingest raw JSON strings
 * stream.ingestRecordOffset("{\"field\": \"value\"}");
 *
 * // Batch ingestion
 * List<MyData> records = ...;
 * Optional<Long> batchOffset = stream.ingestRecordsOffset(records, gson::toJson);
 *
 * stream.close();
 * }</pre>
 *
 * @see ZerobusSdk#createJsonStream(String, String, String)
 */
public class ZerobusJsonStream extends BaseZerobusStream {

  /**
   * A functional interface for JSON serialization.
   *
   * <p>This interface allows using any JSON serialization library (Gson, Jackson, etc.) by passing
   * a serializer function.
   *
   * <p>Example with Gson:
   *
   * <pre>{@code
   * Gson gson = new Gson();
   * stream.ingestRecordOffset(myObject, gson::toJson);
   * }</pre>
   *
   * <p>Example with Jackson:
   *
   * <pre>{@code
   * ObjectMapper mapper = new ObjectMapper();
   * stream.ingestRecordOffset(myObject, obj -> {
   *     try {
   *         return mapper.writeValueAsString(obj);
   *     } catch (JsonProcessingException e) {
   *         throw new RuntimeException(e);
   *     }
   * });
   * }</pre>
   *
   * @param <T> the type of the object to serialize
   */
  @FunctionalInterface
  public interface JsonSerializer<T> {
    /**
     * Serializes an object to a JSON string.
     *
     * @param object the object to serialize
     * @return the JSON string representation
     */
    String serialize(T object);
  }

  /**
   * A functional interface for JSON deserialization.
   *
   * <p>This interface allows using any JSON deserialization library (Gson, Jackson, etc.) by
   * passing a deserializer function.
   *
   * <p>Example with Gson:
   *
   * <pre>{@code
   * Gson gson = new Gson();
   * List<MyData> unacked = stream.getUnackedRecords(json -> gson.fromJson(json, MyData.class));
   * }</pre>
   *
   * @param <T> the type of the object to deserialize to
   */
  @FunctionalInterface
  public interface JsonDeserializer<T> {
    /**
     * Deserializes a JSON string to an object.
     *
     * @param json the JSON string to deserialize
     * @return the deserialized object
     */
    T deserialize(String json);
  }

  // Stream creation parameters (stored for recreateStream)
  private final String clientId;
  private final String clientSecret;

  /** Package-private constructor. Use {@link ZerobusSdk#createJsonStream} to create instances. */
  ZerobusJsonStream(
      long nativeHandle,
      String tableName,
      StreamConfigurationOptions options,
      String clientId,
      String clientSecret) {
    super(nativeHandle, tableName, options, true);
    this.clientId = clientId;
    this.clientSecret = clientSecret;
  }

  /** Returns the client ID used to create this stream. */
  public String getClientId() {
    return clientId;
  }

  /** Returns the client secret used to create this stream. */
  public String getClientSecret() {
    return clientSecret;
  }

  // ==================== Single Record Ingestion ====================

  /**
   * Ingests an object as JSON and returns the offset immediately.
   *
   * <p>This is the main method for JSON ingestion. The object is serialized using the provided
   * serializer function.
   *
   * <p>Example with Gson:
   *
   * <pre>{@code
   * Gson gson = new Gson();
   * stream.ingestRecordOffset(myObject, gson::toJson);
   * }</pre>
   *
   * @param object the object to serialize and ingest
   * @param serializer a function that converts the object to a JSON string
   * @param <T> the type of the object
   * @return the offset ID assigned to this record
   * @throws ZerobusException if the stream is not in a valid state or an error occurs
   */
  public <T> long ingestRecordOffset(T object, JsonSerializer<T> serializer)
      throws ZerobusException {
    ensureOpen();
    String json = serializer.serialize(object);
    return nativeIngestRecordOffset(nativeHandle, json.getBytes(StandardCharsets.UTF_8), true);
  }

  /**
   * Ingests an object as JSON asynchronously without exposing the offset or ack future.
   *
   * <p>This fire-and-forget method serializes the object on the calling thread, then hands the
   * encoded payload off to the native runtime; enqueueing into the stream and applying
   * backpressure happen on the runtime's thread pool. Errors raised after this hand-off (e.g.,
   * the stream fails before the record is acknowledged) are not surfaced to the caller — call
   * {@link #flush()} or {@link #close()} before shutdown when durability matters.
   *
   * <p>Subsequent calls to {@link #flush()}, {@link #close()}, {@link #waitForOffset(long)}, or
   * any offset-returning ingest method on this stream wait for previously-submitted no-wait tasks
   * to complete, so a record submitted here is always observable to those operations.
   *
   * @param object the object to serialize and ingest
   * @param serializer a function that converts the object to a JSON string
   * @param <T> the type of the object
   * @throws ZerobusException if the stream is already closed or serialization fails
   */
  public <T> void ingestRecordNoWait(T object, JsonSerializer<T> serializer)
      throws ZerobusException {
    ensureOpen();
    String json = serializer.serialize(object);
    nativeIngestRecordNoWait(nativeHandle, json.getBytes(StandardCharsets.UTF_8), true);
  }

  /**
   * Ingests a JSON string and returns the offset immediately.
   *
   * <p>Use this method when you already have a JSON string.
   *
   * @param json the JSON string to ingest
   * @return the offset ID assigned to this record
   * @throws ZerobusException if the stream is not in a valid state or an error occurs
   */
  public long ingestRecordOffset(String json) throws ZerobusException {
    ensureOpen();
    return nativeIngestRecordOffset(nativeHandle, json.getBytes(StandardCharsets.UTF_8), true);
  }

  /**
   * Ingests a JSON string asynchronously without exposing the offset or ack future.
   *
   * <p>See {@link #ingestRecordNoWait(Object, JsonSerializer)} for the fire-and-forget contract
   * and the ordering guarantee with respect to subsequent {@link #flush()} / {@link #close()} /
   * offset calls.
   *
   * @param json the JSON string to ingest
   * @throws ZerobusException if the stream is already closed
   */
  public void ingestRecordNoWait(String json) throws ZerobusException {
    ensureOpen();
    nativeIngestRecordNoWait(nativeHandle, json.getBytes(StandardCharsets.UTF_8), true);
  }

  // ==================== Batch Ingestion ====================

  /**
   * Ingests multiple objects as JSON and returns the batch offset.
   *
   * <p>This is the main method for batch JSON ingestion. Each object is serialized using the
   * provided serializer function.
   *
   * @param objects the objects to serialize and ingest
   * @param serializer a function that converts each object to a JSON string
   * @param <T> the type of the objects
   * @return the offset ID for the batch, or empty if the iterable is empty
   * @throws ZerobusException if the stream is not in a valid state or an error occurs
   */
  public <T> Optional<Long> ingestRecordsOffset(Iterable<T> objects, JsonSerializer<T> serializer)
      throws ZerobusException {
    List<byte[]> payloads = new ArrayList<>();
    for (T obj : objects) {
      String json = serializer.serialize(obj);
      payloads.add(json.getBytes(StandardCharsets.UTF_8));
    }
    if (payloads.isEmpty()) {
      return Optional.empty();
    }
    ensureOpen();
    return Optional.of(nativeIngestRecordsOffset(nativeHandle, payloads, true));
  }

  /**
   * Ingests multiple JSON strings and returns the batch offset.
   *
   * <p>Use this method when you already have JSON strings.
   *
   * @param jsonStrings the JSON strings to ingest
   * @return the offset ID for the batch, or empty if the iterable is empty
   * @throws ZerobusException if the stream is not in a valid state or an error occurs
   */
  public Optional<Long> ingestRecordsOffset(Iterable<String> jsonStrings) throws ZerobusException {
    List<byte[]> payloads = new ArrayList<>();
    for (String json : jsonStrings) {
      payloads.add(json.getBytes(StandardCharsets.UTF_8));
    }
    if (payloads.isEmpty()) {
      return Optional.empty();
    }
    ensureOpen();
    return Optional.of(nativeIngestRecordsOffset(nativeHandle, payloads, true));
  }

  /**
   * Ingests multiple objects as JSON asynchronously without exposing the batch offset.
   *
   * <p>See {@link #ingestRecordNoWait(Object, JsonSerializer)} for the fire-and-forget contract
   * and the ordering guarantee with respect to subsequent {@link #flush()} / {@link #close()} /
   * offset calls.
   *
   * @param objects the objects to serialize and ingest
   * @param serializer a function that converts each object to a JSON string
   * @param <T> the type of the objects
   * @throws ZerobusException if the stream is already closed or serialization fails
   */
  public <T> void ingestRecordsNoWait(Iterable<T> objects, JsonSerializer<T> serializer)
      throws ZerobusException {
    List<byte[]> payloads = new ArrayList<>();
    for (T obj : objects) {
      String json = serializer.serialize(obj);
      payloads.add(json.getBytes(StandardCharsets.UTF_8));
    }
    if (payloads.isEmpty()) {
      return;
    }
    ensureOpen();
    nativeIngestRecordsNoWait(nativeHandle, payloads, true);
  }

  /**
   * Ingests multiple JSON strings asynchronously without exposing the batch offset.
   *
   * <p>See {@link #ingestRecordNoWait(Object, JsonSerializer)} for the fire-and-forget contract
   * and the ordering guarantee with respect to subsequent {@link #flush()} / {@link #close()} /
   * offset calls.
   *
   * @param jsonStrings the JSON strings to ingest
   * @throws ZerobusException if the stream is already closed
   */
  public void ingestRecordsNoWait(Iterable<String> jsonStrings) throws ZerobusException {
    List<byte[]> payloads = new ArrayList<>();
    for (String json : jsonStrings) {
      payloads.add(json.getBytes(StandardCharsets.UTF_8));
    }
    if (payloads.isEmpty()) {
      return;
    }
    ensureOpen();
    nativeIngestRecordsNoWait(nativeHandle, payloads, true);
  }

  // ==================== Unacknowledged Records ====================

  /**
   * Returns the unacknowledged records as JSON strings.
   *
   * <p>Use this method when you want to handle deserialization yourself or when you need the raw
   * JSON strings.
   *
   * <p>Note: This method returns cached data after the stream is closed.
   *
   * @return a list of JSON strings
   * @throws ZerobusException if an error occurs
   */
  public List<String> getUnackedRecords() throws ZerobusException {
    List<byte[]> rawRecords;
    if (nativeHandle == 0) {
      rawRecords = getCachedUnackedRecords();
    } else {
      rawRecords = nativeGetUnackedRecords(nativeHandle);
    }
    List<String> result = new ArrayList<>(rawRecords.size());
    for (byte[] bytes : rawRecords) {
      result.add(new String(bytes, StandardCharsets.UTF_8));
    }
    return result;
  }

  /**
   * Returns the unacknowledged records deserialized into objects.
   *
   * <p>Use this method when you need to inspect or process the unacknowledged records as typed
   * objects.
   *
   * <p>Example with Gson:
   *
   * <pre>{@code
   * Gson gson = new Gson();
   * List<MyData> unacked = stream.getUnackedRecords(json -> gson.fromJson(json, MyData.class));
   * for (MyData record : unacked) {
   *     // Process or re-ingest
   * }
   * }</pre>
   *
   * @param deserializer a function that converts JSON strings to objects
   * @param <T> the type of the objects
   * @return a list of deserialized objects
   * @throws ZerobusException if an error occurs
   */
  public <T> List<T> getUnackedRecords(JsonDeserializer<T> deserializer) throws ZerobusException {
    List<String> jsonRecords = getUnackedRecords();
    List<T> result = new ArrayList<>(jsonRecords.size());
    for (String json : jsonRecords) {
      result.add(deserializer.deserialize(json));
    }
    return result;
  }

  /**
   * Returns the unacknowledged batches after stream failure.
   *
   * <p>This method preserves the batch grouping from the original ingestion, which can be useful
   * for re-ingesting records in the same batches.
   *
   * <p>Note: This method returns cached data after the stream is closed.
   *
   * @return a list of unacknowledged batches
   * @throws ZerobusException if an error occurs
   */
  public List<EncodedBatch> getUnackedBatches() throws ZerobusException {
    if (nativeHandle == 0) {
      return getCachedUnackedBatches();
    }
    return nativeGetUnackedBatches(nativeHandle);
  }
}

File: dotnet/src/Databricks.Zerobus/ZerobusJsonStream.cs
using Databricks.Zerobus.Native;
using System.Runtime.InteropServices;
using System.Text;

namespace Databricks.Zerobus;

/// <summary>
/// Stream for ingesting JSON records into a Unity Catalog Delta table.
/// Records are sent as JSON strings and are schema-free — no protobuf
/// compilation is required.
/// </summary>
/// <remarks>
/// JSON streams are ideal for quick starts or when the target schema is dynamic.
/// For production high-throughput workloads, prefer <see cref="ZerobusProtoStream{T}"/>.
/// </remarks>
public sealed class ZerobusJsonStream : BaseZerobusStream
{
    private readonly string _clientId;
    private readonly string _clientSecret;

    /// <summary>
    /// OAuth client ID used for stream authentication.
    /// </summary>
    public string ClientId => _clientId;

    /// <summary>
    /// OAuth client secret used for stream authentication.
    /// </summary>
    public string ClientSecret => _clientSecret;

    internal ZerobusJsonStream(
        IntPtr nativeHandle,
        string tableName,
        StreamConfigurationOptions options,
        string clientId,
        string clientSecret)
        : base(nativeHandle, tableName, options, isJsonMode: true)
    {
        _clientId = clientId;
        _clientSecret = clientSecret;
    }

    // ==================== Single Record Ingestion ====================

    /// <summary>
    /// Ingests a single JSON record string and returns its offset.
    /// </summary>
    /// <param name="json">The JSON record to ingest.</param>
    /// <returns>The offset of the queued record, or -1 on error.</returns>
    public long IngestRecord(string json)
    {
        if (json == null) throw new ArgumentNullException(nameof(json));
        EnsureOpen();

        var bytes = Encoding.UTF8.GetBytes(json);
        GCHandle handle = GCHandle.Alloc(bytes, GCHandleType.Pinned);
        try
        {
            long offset = NativeMethods.zerobus_stream_ingest_json_record(
                NativeHandle,
                handle.AddrOfPinnedObject(),
                (UIntPtr)bytes.Length);

            if (offset < 0)
                throw new ZerobusException("Failed to ingest JSON record.", isRetryable: true);

            return offset;
        }
        finally
        {
            handle.Free();
        }
    }

    // ==================== Batch Ingestion ====================

    /// <summary>
    /// Ingests a batch of JSON record strings and returns the offset of the last record.
    /// Returns null if the input enumerable is empty.
    /// </summary>
    /// <param name="jsonRecords">The JSON records to ingest.</param>
    /// <returns>The offset of the last queued record, or null if empty.</returns>
    public long? IngestRecords(IReadOnlyList<string> jsonRecords)
    {
        if (jsonRecords == null) throw new ArgumentNullException(nameof(jsonRecords));
        if (jsonRecords.Count == 0) return null;
        EnsureOpen();

        var encodedRecords = new byte[jsonRecords.Count][];
        var recordPtrs = new IntPtr[jsonRecords.Count];
        var lengths = new IntPtr[jsonRecords.Count];
        var gcHandles = new GCHandle[jsonRecords.Count];

        try
        {
            for (int i = 0; i < jsonRecords.Count; i++)
            {
                if (jsonRecords[i] == null)
                    throw new ArgumentException($"JSON record at index {i} is null.", nameof(jsonRecords));

                encodedRecords[i] = Encoding.UTF8.GetBytes(jsonRecords[i]);
                gcHandles[i] = GCHandle.Alloc(encodedRecords[i], GCHandleType.Pinned);
                recordPtrs[i] = gcHandles[i].AddrOfPinnedObject();
                lengths[i] = (IntPtr)encodedRecords[i].Length;
            }

            long offset = NativeMethods.zerobus_stream_ingest_json_records(
                NativeHandle,
                recordPtrs,
                lengths,
                (UIntPtr)jsonRecords.Count);

            if (offset == -1)
                throw new ZerobusException("Failed to ingest JSON records batch.", isRetryable: true);
            if (offset == -2)
                return null;

            return offset;
        }
        finally
        {
            foreach (var h in gcHandles) h.Free();
        }
    }

    /// <summary>
    /// Ingests a batch of JSON record strings from an enumerable.
    /// </summary>
    public long? IngestRecords(IEnumerable<string> jsonRecords)
    {
        if (jsonRecords == null) throw new ArgumentNullException(nameof(jsonRecords));
        return IngestRecords(jsonRecords as IReadOnlyList<string> ?? jsonRecords.ToArray());
    }

    // ==================== Unacknowledged Records ====================

    /// <summary>
    /// Returns unacknowledged records as JSON strings.
    /// After the stream is closed, returns cached data.
    /// </summary>
    public IReadOnlyList<string> GetUnackedRecords()
    {
        IReadOnlyList<byte[]> raw;
        if (_disposed != 0 || IsClosed)
            raw = GetCachedUnackedRecords();
        else
            raw = GetNativeUnackedRecords();

        var result = new string[raw.Count];
        for (int i = 0; i < raw.Count; i++)
        {
            result[i] = Encoding.UTF8.GetString(raw[i]);
        }
        return result;
    }

    /// <summary>
    /// Returns unacknowledged records as raw byte arrays.
    /// </summary>
    public IReadOnlyList<byte[]> GetUnackedRecordBytes()
    {
        if (_disposed != 0 || IsClosed)
            return GetCachedUnackedRecords();

        return GetNativeUnackedRecords();
    }
}
